compiler/rustc_const_eval/src/const_eval/eval_queries.rs

```diff
@@ -49,8 +49,12 @@ fn eval_body_using_ecx<'mir, 'tcx, R: InterpretationResult<'tcx>>(
         "Unexpected DefKind: {:?}",
         ecx.tcx.def_kind(cid.instance.def_id())
     );
-    let layout = ecx.layout_of(body.bound_return_ty().instantiate(tcx, cid.instance.args))?;
-    assert!(layout.is_sized());
+
+    let body_ty = body.bound_return_ty().instantiate(tcx, cid.instance.args);
+    let layout = ecx.layout_of(body_ty)?;
+    if !layout.is_sized() {
+        return Err((err_inval!(Layout(LayoutError::Unknown(body_ty)))).into());
```

Comment:
I don't like that I'm returning `LayoutError::Unknown` here. If there's a better error please suggest.

Reply:
I think this is the wrong place to emit an error. Why doesn't this already error in typeck or borrowck?

Reply:
It doesn't error in typeck because the predicate `AA: Sized` evaluates to OK. 

Since `AA` looks like this::

```rust
struct AA {
    pub data: [&usize]
}
```
in order for `AA: Sized` to hold, the nested obligation `[&usize]: Sized` must hold. Unfortunately we fail to generate such nested obligation because  we return `None` from `adt_sized_constraint()`: https://github.com/rust-lang/rust/blob/b14d8b2ef20c64c1002e2c6c724025c3d0846b91/compiler/rustc_ty_utils/src/ty.rs#L90-L96

The reason for that is that the `constraint_ty.references_error()` is true for  `[&usize]`. It is true because of the missing lifetime annotation. As soon as we add a lifetime annotation (e.g. changing `[&usize]` to `[&'static usize]`) the ICE goes away as we now starts to error in typeck.

To fix the issue, I can return `Ty::Error()` from `adt_sized_constraint()` instead of `None` and then return `Ambiguous` in `sized_conditions()`. That seems to fix the ICE. However, I don't believe it is the most appropriate solution.

Reply:
> To fix the issue, I can return `Ty::Error()` from `adt_sized_constraint()` instead of `None` and then return `Ambiguous` in `sized_conditions()`. That seems to fix the ICE. However, I don't believe it is the most appropriate solution.

I like this. The `Ambiguous` is indeed fishy, but compilation is already doomed. Maybe return `Where(vec![ty::Error])`?